api/internal/handler/consumer/consumer_test.go

```diff
@@ -18,230 +18,381 @@
-	"time"
-
-	"github.com/shiningrush/droplet"
-	"github.com/stretchr/testify/assert"
-
+	"context"
+	"fmt"
 	"github.com/apisix/manager-api/internal/core/entity"
-	"github.com/apisix/manager-api/internal/core/storage"
 	"github.com/apisix/manager-api/internal/core/store"
+	"github.com/shiningrush/droplet"
+	"github.com/shiningrush/droplet/data"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"net/http"
+	"testing"
 )
 
-func TestConsumer(t *testing.T) {
-	// init
-	err := storage.InitETCDClient([]string{"127.0.0.1:2379"})
-	assert.Nil(t, err)
-	err = store.InitStores()
-	assert.Nil(t, err)
+func TestHandler_Get(t *testing.T) {
+	tests := []struct {
+		caseDesc   string
+		giveInput  *GetInput
+		giveRet    interface{}
+		giveErr    error
+		wantErr    error
+		wantGetKey string
+		wantRet    interface{}
+	}{
+		{
+			caseDesc:   "normal",
+			giveInput:  &GetInput{Username: "test"},
+			wantGetKey: "test",
+			giveRet:    "hello",
+			wantRet:    "hello",
+		},
+		{
+			caseDesc:   "failed",
+			giveInput:  &GetInput{Username: "failed key"},
+			wantGetKey: "failed key",
+			giveErr:    fmt.Errorf("get failed"),
+			wantErr:    fmt.Errorf("get failed"),
+			wantRet: &data.SpecCodeResponse{
+				StatusCode: http.StatusInternalServerError,
+			},
+		},
+	}
 
-	handler := &Handler{
-		consumerStore: store.GetStore(store.HubKeyConsumer),
+	for _, tc := range tests {
+		getCalled := true
+		mStore := &store.MockInterface{}
+		mStore.On("Get", mock.Anything).Run(func(args mock.Arguments) {
+			getCalled = true
+			assert.Equal(t, tc.wantGetKey, args.Get(0), tc.caseDesc)
+		}).Return(tc.giveRet, tc.giveErr)
+
+		h := Handler{consumerStore: mStore}
+		ctx := droplet.NewContext()
+		ctx.SetInput(tc.giveInput)
+		ret, err := h.Get(ctx)
+		assert.True(t, getCalled, tc.caseDesc)
+		assert.Equal(t, tc.wantRet, ret, tc.caseDesc)
+		assert.Equal(t, tc.wantErr, err, tc.caseDesc)
 	}
-	assert.NotNil(t, handler)
-
-	//create consumer
-	ctx := droplet.NewContext()
-	consumer := &entity.Consumer{}
-	reqBody := `{
-      "username": "jack",
-      "plugins": {
-          "limit-count": {
-              "count": 2,
-              "time_window": 60,
-              "rejected_code": 503,
-              "key": "remote_addr"
-          }
-      },
-    "desc": "test description"
-  }`
-	err = json.Unmarshal([]byte(reqBody), consumer)
-	assert.Nil(t, err)
-	ctx.SetInput(consumer)
-	_, err = handler.Create(ctx)
-	assert.Nil(t, err)
-
-	//create consumer 2
-	consumer2 := &entity.Consumer{}
-	reqBody = `{
-		"username": "pony",
-		"plugins": {
-		  "limit-count": {
-		      "count": 2,
-		      "time_window": 60,
-		      "rejected_code": 503,
-		      "key": "remote_addr"
-		  }
+}
+
+func TestHandler_List(t *testing.T) {
+	tests := []struct {
+		caseDesc  string
+		giveInput *ListInput
+		giveData  []*entity.Consumer
+		giveErr   error
+		wantErr   error
+		wantInput store.ListInput
+		wantRet   interface{}
+	}{
+		{
+			caseDesc: "normal",
```

Comment:
need a clearer description

Reply:
It is a normal case for `List` action, what extra information should we add to it?

Reply:
@ShiningRush  
it contain search and pager, so it should be better if describe the detail.

Reply:
OK